### ISSUE TITLE
[discussion] Pod or job/deployment, that's the question

### ISSUE THREAD

**Issue by gaocegege** [caicloud/kubeflow-controller#71]
Now we have a problem about Kubernetes: Should we use pods or jobs/deployments to create the workers and parameter servers?

Now we use pods although @ScorpioCPH suggests use to use jobs/deployments instead.

## Requirements

- Create workers and parameter servers and allow them to discover each other.
- Allow users to decide the behaviour when worker fails to run.
- Cleanup if all workers finished.
- Delete all workers and parameter servers when users delete the TFJob.

## Pros

- Easy to implement more strategies about workers from the user side.

## Cons

- Hard to manage the lifecycle of the pods, then we have to re-implement some logic in k8s internal controllers.

WDYT @ScorpioCPH @DjangoPeng @mqliang

**Comment by mqliang** [comment]
Personally, I vote for using Pod. Pod is the basic building blocks in k8s, using Pod allows us to better control the whole training process.

**Comment by ScorpioCPH** [comment]
> Easy to implement more strategies about workers from the user side.

Why Pod is `easy to implement more strategies`? Can you give some example?

**Comment by gaocegege** [comment]
I am not familiar with the workflow of tensorflow and I am afraid that I can not give you some examples about the strategies. 

As @DjangoPeng said there are some strategies may be requested from the user side and using pods could help us support these. But I do not know what these strategies are, exactly.

**Comment by mqliang** [comment]
If we use Job, we can only use the functionality/feature provided by Job (Job is a layer above Pod, so it hide some details about Pod). If we want to full control the whole workflow, we should use Pod.

**Comment by ScorpioCPH** [comment]
IMO, `Job/Deployment` is a wrap of `Pod` with some replicas and other control logics, pod-template in `Job/Deployment` is almost the same as `Pod`, I don't know what we can't do with `Job/Deployment` that `Pod` can.

We can set replicas to 1, as we need one service one worker. And after training complete, we can delete these `Jobs/Deployments` to do cleanup.

**Comment by gaocegege** [comment]
They are equal in functionality, IMO. We also could use deployment/job to implement the controller logic, and it is much more easier, in some degree. But if there are some strategies that should be supported and it is hard to use deployments to implement, then pods is the first choice.

So the question is what the strategies are and whether we will meet with those strategies or not, IMO.

I think @DjangoPeng could give us information about it. If job/deployment works well, I think it is better to reuse the logic and no need to manage the lifecycle of pods.

**Comment by ScorpioCPH** [comment]
@mqliang full control is not easy, we need to re-implement some same logic in controller-manager. And we have some issues here: https://github.com/caicloud/kubeflow-controller/issues/36.

**Comment by gaocegege** [comment]
Yeah, full control is not necessary, IMO, if we just need to keep worker and PS up. That is what job controller wants to implement.

**Comment by DjangoPeng** [comment]
Let me introduce some use cases and domain knowledge of TensorFlow.

As we all know, TensorFlow distributed training job includes multiple PS and worker pods. Say 2 PS and 4 worker pods. 

If one worker failed in the training progress, the status of the TFJob is not determinate but up to user code. At least we can divide it into async and sync training mode to discuss individually.

- In async training mode, failure of non-chief worker would not affect the TFJob. So the status should be running. And we don't need to restart the failed worker, because the restart operation may cause TFJob failed if user doesn't implement the `restore model variable from checkpoint file` logic for worker. While if chief worker failed, all workers would fail and the status of TFJob should be failed.
- In sync training mode, failure of any worker would be deadly. 

When it comes to distributed TensorFlow, we always talk about the [data parallelism](https://en.wikipedia.org/wiki/Data_parallelism). The key point for data parallelism is that all worker are doing same computation with different data inputs. It seems like the SIMD in parallel computation. 

In more complicated case, the status of TFJob is up to user code and inputting command argument.  In TensorFlow Cluster, there are two different abstractions: `worker` and `replicas`.

- `worker` is a process in bare metal or pod in Kubernetes cluster. It's the real resource we can use for the TFJob.
- `replicas` is the number of parallelism computation, also known as `task` in TensorFlow.

User could individually set the number of `worker` and `replicas` by TensorFlow Supervisor. 

- In common case, they are equal, so each worker takes a replica/task work in the training progress. TensorFlow doesn't care about the mapping between worker and replica. If we set 4 workers and only launch chief worker, it still works for TensorFlow (but chief worker would take 4 replicas/tasks work). 
- If they are not equal, the Coordinator and Queue would do some logic to keep the distributed training work. It's a little complicated and we can offline sync the details.

In a word, the status of TFJob is not exactly determinate without user specification. In the first stage, we don't need to do any operation if non-chief worker fail.

I'm not a Kubernetes expert, so I have some questions for `Job/Deployment` design.

- I know `Job/Deployment` is a wrapper for `Pod` and we can get the pod info indirectly. But it breaks the hierarchical design of resource, so is it a good design and acceptable convention? 
- If using `Job` for worker, the native successful logic of `Job` is specified number (1 for us) of pods successfully completion. So it doesn't meet our requirement. Maybe we have to do some adoptive works.
- If using `Deployment` for worker, we don't need to restart worker if it's failed. 

FYI. @mqliang @gaocegege-caicloud @ScorpioCPH

**Comment by gaocegege** [comment]
FYI There are some discussions in upstream: https://github.com/tensorflow/k8s/issues/45 And Jeremy think pods may be better.

**Comment by jlewi** [comment]
I added some more information 

https://github.com/tensorflow/k8s/issues/45#issuecomment-355201466

**Comment by ScorpioCPH** [comment]
@DjangoPeng Thanks very much for your detailed explanation.

> But it breaks the hierarchical design of resource, so is it a good design and acceptable convention?

Sorry i'm not very clear about this `breaks the hierarchical design of resource`.

> And we don't need to restart the failed worker, because the restart operation may cause TFJob failed if user doesn't implement the restore model variable from checkpoint file logic for worker. 

> If using Deployment for worker, we don't need to restart worker if it's failed.

We have an option `restartPolicy` in TFJob API, so let user to define the restarting policy maybe better.

> In the first stage, we don't need to do any operation if non-chief worker fail.

If we make a decision for user maybe a little confusing.
